src/runtime/story.rs

```diff
@@ -0,0 +1,192 @@
+use std::{
+    borrow::Borrow,
+    fmt::{self, Write},
+    rc::Rc,
+};
+
+use crate::runtime_graph::RuntimeGraph;
+
+use super::{
+    container::Container,
+    control_command::ControlCommand,
+    value::Value,
+    variable::{ReadCount, VariableReference},
+    RuntimeObject,
+};
+
+#[derive(Debug)]
+pub struct Story<Output> {
+    output_text: Output,
+    glue: bool,
+    cursors: Vec<(Rc<Container>, usize)>,
+}
+
+impl<Output> Story<Output> {
+    fn new<Graph>(graph: Graph, output_text: Output) -> Self
+    where
+        Graph: Borrow<RuntimeGraph>,
+    {
+        let cursors = vec![(graph.borrow().root_container.clone(), 0)];
+        Self {
+            output_text,
+            glue: false,
+            cursors,
+        }
+    }
+
+    pub fn step(&mut self) -> Result<(), ()>
+    where
+        Output: Write,
+    {
+        let object = self.peek_cursor().ok_or(())?;
+        if self.execute(object.clone()) {
+            self.advance_cursor()?;
+        }
+        Ok(())
+    }
+
+    fn execute(&mut self, object: RuntimeObject) -> bool
```

Comment:
I might be pre-optimizing here but it seems like we don't need Execute to take ownership of the runtime object, especially if it's going to result in a clone on every step

Reply:
`step` function has an exclusive `&mut self`, and then borrows a `&RuntimeObject` from it. `execute` takes both the `RuntimeObject` and `&mut self`, so the borrow checker won't allow `&object` as is.`execute` isn't allowed to use both the exclusive borrow and the object borrowed at the same time. Solutions:
- Clone (this seemed natural at the time, but now I think about it, `RuntimeObject` can contain a long un-interned string or path)
- Split up the borrows, so `execute` only borrows part of the story struct, disjoint with the cursors (messy, requires changing the pattern to something like `self.cursors.advance()` and `self.rest.execute(object_borrowed_from_cursors)`)
- `Rc` somehow, I haven't fully thought this through

Reply:
- Have `execute` do the borrow itself from its `&mut self`, in which case may as well inline the `execute` method.

Reply:
Resolving for now, see #18